Add support for VRRP and IGMP in host fw

aditighag opened the pull request:
This PR adds support for VRRP and IGMP protocols in host firewall. 

### Commit highlights

- Policy api/engine: The existing `toPorts` field is extended to support these protocols. As the protocols don't have transport ports, the existing validation check around empty ports is relaxed. Users can explicitly whitelist/deny these protocols using the `toPorts` rules, and these rules can be combined with the existing L3 rules. 
- BPF datapath: Conntrack will now pass packets from these protocols that would previously be dropped with `DROP_CT_UNKNOWN_PROTO` error. This is gated behind a config flag so that it won't break existing allow-all policies. The decision to allow/deny such traffic is delegated to user-defined policies. Also, the existing behavior of allowing reply traffic based on conntrack state is preserved for VRRP and IGMP protocols as well.

### Additional notes

- Policy api: Earlier version of this PR introduced a new field `IPProtocols` that would allow users to specify such IP protocols without transport ports using their IANA assigned numbers. This field was restricted to CCNP only using CEL rules. However, based on discussions with @jrajahalme and @squeed, we decided to reuse the existing `toPorts` field instead as the protocols don't need header fields matching similar to ICMP, and so the current use cases can be satisfied without introducing the new field. Note that the policy engine already supports policy matching with wildcard port field.
- Policy enforcement for hosts and pods: Support for VRRP and IGMP protocols has been extended to both host and pod traffic, leveraging the shared policy enforcement and conntrack logic common to bpf_host and bpf_lxc programs. This PR, however, focuses on host firewall in terms of testing. Filed issue for bpf_lxc tests - https://github.com/cilium/cilium/issues/40229. 

Results from manual testing:
```
$ cat lock-down-hw.yaml
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "lock-down-ingress-worker-node"
spec:
  description: "Allow selected traffic on egress of worker nodes"
  nodeSelector:
    matchLabels:
      node-access: policy
  egress:
  - toPorts:
    - ports:
      - port: "22"
        protocol: TCP
      - port: "443"
        protocol: TCP
      - port: "6443"
        protocol: TCP
      - port: "2379"
        protocol: TCP
      - port: "4240"
        protocol: TCP
      - port: "8472"
        protocol: UDP
      - port: "0"
        protocol: VRRP
```

- Apply a host fw policy that allows L4 traffic on some ports but not VRRP. Traffic is denied with no match.
`Policy verdict log: flow 0x0 local EP ID 559, remote ID world, proto 112, egress, action deny, auth: disabled, match none, 172.19.0.4 -> 224.0.0.18`
- Extend the `toPorts` rule to explicitly allow VRRP. Traffic is allowed.
- Explicitly deny VRRP traffic with toPorts rule. Traffic is denied with proto match.
`Policy verdict log: flow 0x0 local EP ID 559, remote ID world, proto 112, egress, action deny, auth: disabled, match Proto-Only, 172.19.0.4 -> 224.0.0.18`
- Extend deny rules with to/forCIDR rules for VRRP. Traffic is denied via L3 match.
`Policy verdict log: flow 0x0 local EP ID 559, remote ID 16777217, proto 112, egress, action deny, auth: disabled, match L3-Only, 172.19.0.4 -> 224.0.0.18
`

Follow-up: E2E CI test and Hubble proto extensions

Fixes: #31502 
Fixes: #18347 

```release-note
Add support for VRRP and IGMP protocols in host firewall
```

aditighag committed:
policy/api: Extend toPorts with additional protocols

This commit extends the network policy spec to allow/deny
protocols such as VRRP and IGMP that don't have transport
ports. Note that if protocol is set to ANY and port is omitted,
the policy may permit any *L4* protocols (TCP, UDP, and SCTP).

Signed-off-by: Aditi Ghag <aditi@cilium.io>

squeed commented on pkg/policy/api/ip_protocol.go:
```diff
@@ -0,0 +1,38 @@
+
+// IPProtocols is a list of IP protocols that are allowed. This field is intended for protocols
+// that operate directly over IP without transport-layer ports (unlike TCP or UDP).
+type IPProtocols []IPProtocol
```
Maybe this could implement `pkg/policy/api.PortsIterator`? Then we could treat it the same as port+proto and icmp?

squeed commented on pkg/policy/api/egress.go:
```diff
@@ -104,6 +104,19 @@ type EgressCommonRule struct {
+	// by their IP protocol number (as defined in the IP header 'Protocol' field).
+	//
+	// +kubebuilder:validation:Optional
+	IPProtocols IPProtocols `json:"ipProtocols,omitempty"`
```
I'm torn between adding a new field for maximum clarity, and relaxing the restriction that PortRule have a port number. Since we support port ranges, we already can express "allow all TCP packets" (albeit awkwardly). Why did you decide to add a new field?

squeed commented on pkg/policy/api/rule_validation.go:
```diff
@@ -640,16 +640,12 @@ func (pr *PortDenyRule) sanitize() error {
 }
 
 func (pp *PortProtocol) sanitize(hasDNSRules bool) (isZero bool, err error) {
-	if pp.Port == "" {
```
Should we keep this restriction here if `!option.Config.EnableBPFConntrackUnknownProtocols`? Or are VRRP / IGMP always allowed?

squeed commented:
Wow, this isn't a very big change at all. Nice!

squeed commented:
Thinking a bit more about the ANY problem. I'd l like to make sure the API is as consistent as possible.

Right now, it's not allowed to specify ANY without a port / port range. I think we should keep it that way, and document that ANY means "any protocol that has an l4 port (tcp / udp / sctp)". We can easily enforce this in CEL.

We should also enforce that vrrp and igmp cannot have port numbers. Hopefully this is doable in CEL.

Finally, I understand why this is behind a feature gate right now: if a pod has a L3-only policy, this will allow previously-denied traffic. However, the chances of a pod actually *listening* to vrrp are essentially nil, so I think we can ultimately remove the feature gate. We did the same thing when we added SCTP support to conntrack, so there is precedent for this. As long as the chance is communicated, I think it's OK.

Thanks to the `enableDefaultDeny: false` functionality, end-users can block IGMP / VRRP by default trivially.

aditighag commented:
> Thinking a bit more about the ANY problem. I'd l like to make sure the API is as consistent as possible.
> 
> Right now, it's not allowed to specify ANY without a port / port range. I think we should keep it that way, and document that ANY means "any protocol that has an l4 port (tcp / udp / sctp)". We can easily enforce this in CEL.
> 

We're aligned on the semantics of `ANY` when it comes to protocols that use transport ports. However, I'm on the fence about rejecting rules that specify an empty port combined with protocol `ANY`.
On one hand, such a rule might seem ambiguous or overly broad. On the other hand, from a UX perspective, it could serve as a convenient way for users to allow only transport protocols like TCP, UDP, and SCTP -- without having to explicitly deny every other non-port-based protocol like IGMP or VRRP. That said, the feature flag `enable-unknown-protocols` does provide this behavior by default, although not with policy ingress/egress semantics.

So the question is: should we allow `port: 0` and `protocol: ANY` as a user-friendly shorthand to mean "only TCP/UDP/SCTP, and not other protocols without transport ports"?

> We should also enforce that vrrp and igmp cannot have port numbers. Hopefully this is doable in CEL.
> 

Good point! Agreed! 💯 

> Finally, I understand why this is behind a feature gate right now: if a pod has a L3-only policy, this will allow previously-denied traffic. However, the chances of a pod actually _listening_ to vrrp are essentially nil, so I think we can ultimately remove the feature gate. We did the same thing when we added SCTP support to conntrack, so there is precedent for this. As long as the change is communicated, I think it's OK.
> Thanks to the `enableDefaultDeny: false` functionality, end-users can block IGMP / VRRP by default trivially.


Users could also apply an allow-all host fw policy, no? +1 for removing the feature gate in future releases.

joestringer commented:
Based on the active discussions including on API and draft status, and given the feature freeze date is imminent, I think it's too late to add this as a release blocker for Cilium v1.18.

aditighag commented:
/test

squeed commented:
> On one hand, such a rule might seem ambiguous or overly broad. On the other hand, from a UX perspective, it could serve as a convenient way for users to allow only transport protocols like TCP, UDP, and SCTP -- without having to explicitly deny every other non-port-based protocol like IGMP or VRRP.

Oh, that totally makes sense. If we allow for protocol specification, and we support protocol-level granularity in the datapath, there's no compelling reason not to expose this knob.

So, +1 for just allowing protocol without port in all cases.

aditighag commented:
/test

aditighag committed:
pke/policy: Support extended protocols

Extend policy support to include protocols such
as VRRP and IGMP, which do not use transport-layer ports.

For these protocols, the port field in toPorts rules
will be either omitted or explicitly set to 0.
Policy matching logic treats the port field as a
wildcard in such cases, and matches on the
protocol specified in the policy key.

To enable this, relax port validation to allow toPorts
rules without a port field when applicable.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf, config: Enable extended IP protocols traffic

Pass traffic with IP protocols like
VRRP and IGMP in conntrack datapath. The traffic 
will be allowed/denied per user-defined policies.

It's guarded behind a configuration flag to
avoid allowing this traffic by default, especially
for users who may have previously applied allow-all
policies.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf/tests: Add tests for extended protocols in host fw

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
docs: Document VRRP + IGMP support in host firewall

Signed-off-by: Aditi Ghag <aditi@cilium.io>

jrajahalme commented:
/test

aditighag commented:
/test

jrajahalme commented on pkg/policy/api/l4.go:
```diff
@@ -49,15 +49,17 @@ type PortProtocol struct {
-	// Protocol is the L4 protocol. If omitted or empty, any protocol
-	// matches. Accepted values: "TCP", "UDP", "SCTP", "ANY"
+	// Protocol is the L4 protocol. If omitted or empty, any protocols
+	// with transport ports (TCP, UDP, SCTP) match.
```
This is a bit wrong. "If omitted or empty, any protocol matches." is still correct. So this should be:

> Protocol is the L4 protocol. If omitted or empty, any protocol matches. "ANY" matches any protocol with L4 ports ("TCP", "UDP", "SCTP").

jrajahalme reviewed (commented):
Small nits.

aditighag commented on bpf/lib/conntrack.h:
```diff
@@ -17,6 +17,8 @@
 #include "signal.h"
 #include "ipfrag.h"
 
+DECLARE_CONFIG(bool, enable_unknown_protocols, "Pass traffic with unknown protocols")
```
I don't have a strong preference, but it seems a bit out of place to talk about policy enforcement in the conntrack related processing code. I could reword the user facing flag to say that "Pass traffic with unknown protocols in the datapath, as policy enforcement will handle validation."

ti-mo replied:
A few observations on the naming and placement of the config:

- It's currently only used in conntrack, but it's also bleeding into the xdp, overlay and wireguard objects where the var is currently left to its default value of `false`. Could we avoid this somehow? This is going to confuse the heck out of whomever tries to use the variable in one of those objects in the future, since they'll have to add the necessary glue code to the respective loader entry points of those objects to get it to change to `true`. These three objects may not need to import `conntrack.h` to begin with, so it may be worth investigating why this is the case and dropping the import if possible. That way, we don't end up with unused fields in config objects.
- It seems there are only hostfw-related tests, may we assume this variable is hostfw-specific? If so, it could be prefixed or called out in the description. 
- Is the use of the variable expected to increase in the future? If so, it might make sense to decouple it from conntrack and move it closer to hostfw code since neither the config name nor description mention this explicitly. (and the description is still rather vague)
- `traffic with unknown protocols is supported in datapath` still feels vague. Does `Forward traffic with unknown protocols to and from Pods` still make sense? The original gives the impression unknown protocols are universally supported, but based on the added datapath code I think this is only the case for delivering locally to Pods (e.g. no forwarding to other nodes for Services), or did I read that incorrectly? In any case, it doesn't hurt being specific, it can always be expanded later.

aditighag replied:
>Does Forward traffic with unknown protocols to and from Pods still make sense?

The code is applicable for both pods and hosts. 

>These three objects may not need to import conntrack.h to begin with, so it may be worth investigating why this is the case and dropping the import if possible.

I'll try it moving it to `LocalNodeConfiguration` per your suggestion.

squeed replied:
Unless I'm mistaken, and setting this truly enables all L4 protocols.

aditighag replied:
@ti-mo The variable is in correct place, but I missed adding it to xdp, overlay and wireguard configuration objects. As I mentioned earlier, the programs either directly import `conntrack.h` or via `lb.h`.

aditighag commented:
All tests passed. 

The latest push has code comment changes only.

aditighag commented:
/test

aditighag commented:
/test

yeskiy commented:
Oftopic:
Hey! I wish you all the best in finishing this pr.

Bit out of curiosity: 
Can we call a flag `-enable-unknown-protocols` a workaround for #17877?

aditighag commented on bpf/lib/conntrack.h:
```diff
@@ -821,6 +830,12 @@ ct_extract_ports4(struct __ctx_buff *ctx, struct iphdr *ip4, fraginfo_t fraginfo
 					  dir, &tuple->dport);
 	default:
+		/* Traffic is allowed/dropped based on user-defined policies. */
+		if (CONFIG(enable_unknown_protocols)) {
```
>I don't quite like the naming of this config.

I'm planning to change the name to `EnableExtendedProtocols`. 

>We definitely don't treat IGMP and VRRP as unknown protocols, when we match on nexthdr.

What do you mean by "match on nexthdr"? Currently, packets with these protocols is dropped with `CT_UNKNOWN_PROTO` error. 

>Does setting EnableUnknownProtocols change the semantics of ANY to include IGMP and VRRP?

No, it doesn't. 

>If we want to keep the semantics of ANY, i.e. not include IGMP and VRRP, we can put them into a separate map and do two lookups (not sure if it's worth it, though).

That seems like an overkill. 

>I'd also rather be more explicit in this switch-case, and match on IGMP and VRRP explicitly, setting ports to 0 only for these protocols

It's an option, but users may want to enable other IP protocols without any transport ports. Is it worth extending the switch case every time we need to support a new protocol in the datapath? The goal is to delegate allow/deny decisions to policy based enforcement. Btw we already have a precedent for passing through such traffic - https://github.com/cilium/cilium/pull/31820/commits/60b233b6ea46e41d58a80a6604b7ef0ffcaf4468.

aditighag commented on pkg/u8proto/u8proto.go:
```diff
@@ -16,18 +16,22 @@ const (
 	// ANY represents all protocols.
```
I'll update the comment. Thanks!

squeed commented on pkg/policy/api/l4.go:
```diff
@@ -40,7 +42,7 @@ type PortProtocol struct {
 	//
 	// +kubebuilder:validation:Pattern=`^(6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5][0-9]{4}|[0-9]{1,4})|([a-zA-Z0-9]-?)*[a-zA-Z](-?[a-zA-Z0-9])*$`
-	Port string `json:"port"`
+	Port string `json:"port,omitempty"`
```
Should we add validation that `port` is not allowed when VRRP and IGMP are set?

aditighag replied:
Yup, here is the validation - https://github.com/cilium/cilium/pull/39872/commits/d026da14ec0addf3e50d7a6f3c88607e37ebb74f#diff-cecfad25fc16b5cc6406ceabe9950e8ed4889e5a7ec9ab8d1f854ee5305e907fR654-R657.

squeed commented on pkg/policy/api/rule_validation.go:
```diff
@@ -641,15 +641,20 @@ func (pr *PortDenyRule) sanitize() error {
 func (pp *PortProtocol) sanitize(hasDNSRules bool) (isZero bool, err error) {
 	if pp.Port == "" {
-		return isZero, errors.New("Port must be specified")
+		if !option.Config.EnableUnknownProtocols {
```
as discussed, I think it's OK to allow empty ports for all protocols. The PolicyMap supports it just fine.

aditighag replied:
I added the conditional validation per your review comment - https://github.com/cilium/cilium/pull/39872#discussion_r2144511283.

aditighag commented:
/test

aditighag committed:
bpf/tests: Add helper utils

This is a preparatory commit to
add helper functions for adding
L4 ingress policy entries to
match on protocol bits.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented:
/test

dylandreimerink reviewed (approved):
Changes look good for my codeowners

qmonnet commented on Documentation/security/policy/language.rst:
```diff
@@ -1562,6 +1562,27 @@ the label ``type=ingress-worker`` on TCP ports 22, 6443 (kube-apiserver), 2379
+the label ``type=egress-worker`` on TCP ports 22, 6443/443 (kube-apiserver), 2379
+(etcd), and 4240 (health checks), UDP port 8472 (VXLAN), and traffic with VRRP protocol.
+
+.. only:: html
+
+   .. tabs::
+     .. group-tab:: k8s YAML
+
+        .. literalinclude:: ../../../examples/policies/host/allow-extended-protocols.yaml
+          :language: yaml
+
+.. only:: epub or latex
+
+        .. literalinclude:: ../../../examples/policies/host/allow-extended-protocols.yaml
+          :language: yaml
```
Please take a look at https://github.com/cilium/cilium/pull/40212/files#diff-a0b27a8b1488de6bbbb7e5f8282627bcff6d32826f2aa2341d8af07387df34b7, it's probably worth ditching the different format conditions here to align with Casey's clean-up.

qmonnet replied:
https://github.com/cilium/cilium/pull/40212 has just be merged, please rebase.

qmonnet reviewed (approved):
I don't have the time to look at the changes in detail and only covered doc changes, sorry - Your changes look good, but Casey has an open PR that cleans up the docs and you may want to adjust accordingly, see below.

pchaigno commented on bpf/tests/hostfw_extended_protocols.c:
```diff
@@ -0,0 +1,336 @@
+}
+
+/* Send an IGMP packet from the IGMP destination to host. Packet is allowed based on
+ * conntrack entry (no ingress policy).
```
Aren't the different tests supposed to stay independent? As in, the second should be able to run on its own.

aditighag replied:
I'm not aware of such a requirement. We seem to have a few instances of dependent tests - https://github.com/cilium/cilium/blob/main/bpf/tests/hostfw_host_iptables.c and https://github.com/cilium/cilium/blob/main/bpf/tests/inter_cluster_snat_clusterip_backend_lxc.c.

pchaigno commented on pkg/policy/api/l4.go:
```diff
@@ -49,15 +49,17 @@ type PortProtocol struct {
-	// Protocol is the L4 protocol. If omitted or empty, any protocol
-	// matches. Accepted values: "TCP", "UDP", "SCTP", "ANY"
+	// Protocol is the L4 protocol. If "ANY", omitted or empty, any protocols
+	// with transport ports (TCP, UDP, SCTP) match.
```
Shouldn't ANY also match VRRP and IGMP?

aditighag replied:
The policy engine adds explicit entries only for TCP, UDP, and SCTP when the protocol is set (or defaults) to `ANY` - https://github.com/cilium/cilium/blob/main/pkg/policy/rule.go#L613-L630. 

This PR doesn't change the semantics for ANY. See more context in one of my earlier comments - https://github.com/cilium/cilium/pull/39872#issuecomment-2970758191.

aditighag replied:
I believe this note is descriptive enough - 

```
If "ANY", omitted or empty, any protocols
	// with transport ports (TCP, UDP, SCTP) match.
```

I'm not sure if it's worth mentioning every IP protocol in case we extend the list for other protocols in future. Happy to extend the note if users report confusion around the `ANY` usage.

pchaigno commented on pkg/policy/api/rule_validation.go:
```diff
@@ -641,15 +641,20 @@ func (pr *PortDenyRule) sanitize() error {
+	} else if pp.Port != "" {
+		if pp.Port != "0" && (pp.Protocol == ProtoVRRP || pp.Protocol == ProtoIGMP) {
+			return isZero, errors.New("port must be empty or 0")
+		}
```
When `EnableExtendedIPProtocols` is true, an unspecified `Port` with UDP or TCP `Protocol` is allowed when it used to be rejected before this commit. Is that expected?

aditighag replied:
Yes, see - https://github.com/cilium/cilium/pull/39872#issuecomment-2977754154 and https://github.com/cilium/cilium/pull/39872#discussion_r2144511283. 
I'm not sure if this warrants a note in the upgrade guide though.

aditighag replied:
Sure, I'll add as a follow-up.

pchaigno reviewed (commented):
Thanks for the neat pull request! Very easy to review :pray: 

I have a couple questions/concerns below.

pchaigno reviewed (approved):
Thanks for the pointers to previous discussions! LGTM!

I think my latest comments can be addressed as follow ups since they're more about documentation.
